### PR TITLE
Add compile step that bundles and compresses JS files

### PR DESCRIPTION
**Pull request by timjb**
Also, manage dependencies on third-party JS libraries using NPM. This addresses the second point in #683.

Todo:

* [x] improve JS code using ES2015 features and switch to TypeScript

### CODE REVIEW

File: .ghc.environment.x86_64-darwin-8.2.1 [hunk 1]
```diff
@@ -0,0 +1,60 @@
+-- This is a GHC environment file written by cabal. This means you can
```
Comment (alexbiehl) [inline .ghc.environment.x86_64-darwin-8.2.1]: Please remove this file from the diff. We should probably put .ghc.environment.* to .gitignore.

File: haddock-api/src/Haddock/Utils.hs [hunk 2]
```diff
@@ -326,16 +326,10 @@ makeAnchorId (f:r) = escape isAlpha f ++ concatMap (escape isLegal) r
 
 
 jsFile :: String
```
Comment (alexbiehl) [inline haddock-api/src/Haddock/Utils.hs]: Could you rename jsFile to something more meaningful? Now that we have several JavaScript includes this derserves a less generic name I think. How about haddockJsFile?

Comment (timjb) [reply haddock-api/src/Haddock/Utils.hs]: Fixed.

File: .ghc.environment.x86_64-darwin-8.2.1 (hunk 1 above)
Comment (timjb) [reply .ghc.environment.x86_64-darwin-8.2.1]: Fixed.

**Comment by alexbiehl** [comment]
This patch changes the QuickJump interface. We must not forget to change hackage accordingly: 

It needs to know how to load the new `quick-jump.js` on package pages. Knowing QuickJump would undergo some changes I implemented a versioning scheme in https://github.com/haskell/hackage-server/commit/51d86a372cc28817501cbe07d942e5f3f6ee7afb#diff-bed07c0e9d3155a8916521a5bd83a710R258. Haddock generates a `meta.json` which hackage inspects and chooses which version of QuickJump it has to deal with. 

Currently it decides on the `haddock_version` field  from `meta.json` but I think we should introduce a `quickjump_version` field to `meta.json`. And just like the way interface files are versioned (c.f. https://github.com/haskell/haddock/blob/master/haddock-api/src/Haddock/InterfaceFile.hs#L86) we increment the version everytime a breaking change to the QuickJump interface happens. This way we are decoupled from the haddock version.

**Comment by timjb** [comment]
I don't see the need for a separate new version number. I believe that we can keep the QuickJump interface stable going forward. For instance, this pull request doesn't change the `quickNav.init` interface, it only changes the organization of the JS files (one minified, bundled file instead of three). If we ever need to pass an additional argument to `quickNav.init` we can just do so, without breaking backwards compatibility.

**Comment by alexbiehl** [comment]
Looks good to me. Just one more thing: Please separate the QuickJump css from haddock css.

**Comment by alexbiehl** [comment]
I will prepare the corresponding commit to hackage.

**Comment by timjb** [comment]
> Just one more thing: Please separate the QuickJump css from haddock css.

Done. The layout bug should be gone when `quick-jump.css` is used together with the Hackage CSS.

**Comment by alexbiehl** [comment]
Great! Looks good. Will merge as soon as travis is green.

**Comment by alexbiehl** [comment]
Dang. I think you forgot to update the tests.

**Comment by timjb** [comment]
> Dang. I think you forgot to update the tests.

Yep. I've sent an updated patch.

**Pull request merged**